tests/nodes/utils.ts

```diff
@@ -10,11 +10,14 @@ import { IdInternal } from '@matrixai/id';
  * 2^i <= distance (from current node) < 2^(i+1)
  * Therefore, generatedNodeId = 2^i XOR nodeId.
  */
-function generateNodeIdForBucket(nodeId: NodeId, bucketIndex: number): NodeId {
+function generateNodeIdForBucket(
+  nodeId: NodeId,
+  bucketIndex: number,
+  alt: number = 0,
+): NodeId {
   const lowerBoundDistance = BigInt(2) ** BigInt(bucketIndex);
   const bufferId = Buffer.from(nodeId.toBuffer());
-  // Console.log(bufferId);
-  const bufferDistance = bigIntToBuffer(lowerBoundDistance);
+  const bufferDistance = bigIntToBuffer(lowerBoundDistance + BigInt(alt));
   // Console.log(bufferDistance);
   // Console.log('Distance buffer:', bufferDistance);
   // console.log('Node ID buffer:', bufferId);
```

Comment:
What does the `alt` parameter do? Can you add an explanation into the comments there.

There's also some `console.log` being used here that should be removed.

Reply:
Note to self, the reason to use `BigInt` is because the key size is 256 bits, so we have numbers up to `2^256`. If we ever change our key size, then this keyspace size will change, so that's why `BigInt` is necessary.

Reply:
Added description for alt into the inline documentation. Removed the commented console.logs

Reply:
Ok I just read the document, I need you to explain to me how exactly does `alt` specify a deterministic alternative node ID, that still doesn't make sense to me.

Secondly.. this seems unnecessary:

```
  const bufferId = Buffer.from(nodeId.toBuffer());
```

Just use `nodeId.toBuffer();`

Also the `bigIntToHexBuffer`, please provide a type for that. That's a `Uint8Array`, and what is a "hex buffer". Is it a buffer of hexadecimal digits? In that case, that's just a string buffer. Or is a buffer of something else? Because looking at the utility function it appears to be doing something else.

Whenever you do these operations, inline comments should be done. For example adding leading `0` should be a `// Padding with leading 0s`.

Also the function seems to be converting it to a binary number. It's not really a hex buffer is it? It's just a binary version of an arbitrary sized number. Normally we would write a 32 bit number, 64 bit number. But big int is arbitrary sized. So you should call this `bigIntToBuffer` and do you want to use Node buffers or `Uint8Array`? If `Uint8Array`, it might be better to call it `bigIntToUint8Array`, and this seems like a useful utility we can stick in `utils.ts`.

According to https://www.npmjs.com/package/bigint-buffer, docblock comments about whether this is Big Endian or Little Endian would also be helpful.

Reply:
> Ok I just read the document, I need you to explain to me how exactly does alt specify a deterministic alternative node ID, that still doesn't make sense to me.

I've taken a look, and the following is my understanding of it:

Remember with Kademlia, a node is stored in bucket `i` where `2^i <= distance (from current node) < 2^(i+1)`. And also remember that `distance = nodeId1 XOR nodeId2`.

So this line:

```ts
const lowerBoundDistance = BigInt(2) ** BigInt(bucketIndex);
```

This calculates that `2^i` value - so its our lower bound of the distance. Therefore, assuming we have some known `nodeId1`, then we can calculate a `nodeId2` by performing `nodeId2 = lowerBoundDistance XOR nodeId1`. `nodeId2` would refer to the node ID with that `lowerBoundDistance` from `nodeId1`.

So when we do:

```ts
const bufferDistance = bigIntToBuffer(lowerBoundDistance + BigInt(alt));
```

All we're doing is adding some integer value to this lower bound, such that we can calculate a different node ID in the same bucket. e.g.

```
nodeId3 = (lowerBoundDistance + 1) XOR nodeId1
nodeId4 = (lowerBoundDistance + 2) XOR nodeId1
nodeId5 = (lowerBoundDistance + 3) XOR nodeId1
...
```